### PR TITLE
Sort filenames when reading parquet to ensure consistent schema

### LINKED ISSUE #5779: Different schemas when inferring from local system in different OS
### Describe the bug

When inferring a [schema](https://github.com/apache/arrow-datafusion/blob/c8a3d589889dd1e67047de89db8b4ff56f90f04c/datafusion/core/src/datasource/listing/table.rs#L431), the [list_all_files](https://github.com/apache/arrow-datafusion/blob/c8a3d589889dd1e67047de89db8b4ff56f90f04c/datafusion/core/src/datasource/listing/url.rs#L143) uses an object store to list the files. No sorting is passed. 
When the object store is a LocalFileSystem, there isn't an insurance of any file sorting (the return list of a macOs has a different sort of windows). This means that the inferred schema can be different for the same set of files.

We contact the object store (https://github.com/apache/arrow-rs/issues/3975) that point it out that the solution should be implemented in the caller of the method, applying a sort of any type, to maintain consistency between file systems.

### To Reproduce

Having two parquet files in the filesystem with the schema:
- file1.parquet
```
{
  "type" : "record",
  "name" : "root",
  "fields" : [ {
    "name" : "year",
    "type" : [ "null", "int" ],
    "default" : null
  }, {
    "name" : "description",
    "type" : [ "null", "string" ],
    "default" : null
  }, {
    "name" : "code",
    "type" : [ "null", "long" ],
    "default" : null
  } ]
}
```
- file3.parquet
```
{
  "type" : "record",
  "name" : "root",
  "fields" : [ {
    "name" : "description",
    "type" : [ "null", "string" ],
    "default" : null
  }, {
    "name" : "code",
    "type" : [ "null", "long" ],
    "default" : null
  }, {
    "name" : "year",
    "type" : [ "null", "int" ],
    "default" : null
  } ]
}
```

and executing:
```
#[tokio::test]
async fn infer_schema() {
    let path =  ListingTableUrl::parse("./files").unwrap();
    let ctx = SessionContext::new();
    let state = ctx.state();
    let options = ListingOptions::new(Arc::new(ParquetFormat::default()));

    let schema = options.infer_schema(&state, &path).await.unwrap();

    schema.fields.iter().for_each(|field|  println!("{0}", field.name()));
}
```
the result in macOs Ventura:
```
description
code
year
```
the first file pickup was the file3.parquet
and using windows
```
year
code
description
```
the first file pickup was the file1.parquet

### Expected behavior

The same schema independently the OS where the code is run. A sort should be forced or at least given the possibility of passing a sort function

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by thomas-k-cameron**
# Which issue does this PR close?
Closes #5779.

# Rationale for this change

Sorting is passed to ensure that files returned is in the same order when the OS is different.

# What changes are included in this PR?

This PR adds a lines of code to `FileFormat::infer_schema` to `ParquetFormat`; It sorts parquet file by it's location.  

On `test_util::store_parquet` is changed to sort temp file by it's path and and write batches in the sorted order. 

This change was introduced as it made the some tests, such as `datasource::file_format::parquet::tests::read_merged_batches` to fail.

It appears that these tests assumed that `infer_schema` processed the files in the same order that  `store_parquet`'s returning value files.

# Are these changes tested?
Yes. This PR introduces new test case `async fn issue_5779()`.

# Are there any user-facing changes?
The order of `Fields` may change.

### CODE REVIEW

**Comment by alamb** [comment]
Thanks @thomas-k-cameron  -- given my reading of https://github.com/apache/arrow-datafusion/issues/5779 it seems like perhaps a more direct solution might be to sort the list of files so they are always processed in the same order

Perhaps somehow figure out how to sort the `objects`  by the [`location`](https://docs.rs/object_store/0.6.1/object_store/struct.ObjectMeta.html#structfield.location) field, for example?

```rust
        let schemas: Vec<_> = futures::stream::iter(objects)
            .map(|object| fetch_schema(store.as_ref(), object, self.metadata_size_hint))
            .boxed() // Workaround https://github.com/rust-lang/rust/issues/64552
            .buffered(SCHEMA_INFERENCE_CONCURRENCY)
            .try_collect()
            .await?;
```

**Comment by thomas-k-cameron** [comment]
Thank you for your advice.

I couldn't figure out what I was getting it wrong.   
Let me try it out.

**Comment by thomas-k-cameron** [comment]
Let me give you an update.

I tried to sort by location, but it hasn't been working.
I end up getting error from tests for fetching statistics.

It seems like this is coming from how parquet files for tests are created with random names.

I have few other ideas. I will see if it works out.

**Comment by thomas-k-cameron** [comment]
It's been a while but I believe it is ready for review.

I updated the PR comment to reflect my changes.

File: datafusion/core/src/datasource/file_format/parquet.rs [hunk 1]
```diff
@@ -176,13 +186,26 @@ impl FileFormat for ParquetFormat {
...
+                    object,
+                    self.metadata_size_hint,
+                )
+            })
             .boxed() // Workaround https://github.com/rust-lang/rust/issues/64552
             .buffered(state.config_options().execution.meta_fetch_concurrency)
             .try_collect()
             .await?;
 
+        schemas.sort_by(|(location1, _), (location2, _)| location1.cmp(location2));
```
Comment (alamb) [inline datafusion/core/src/datasource/file_format/parquet.rs:202]:
```suggestion
        // Schema inference adds fields based the order they are seen
        // which depends on the order the files are processed. For some 
        // object stores (like local file systems) the order returned from list
        // is not deterministic. Thus, to ensure deterministic schema inference 
        // sort the files first. 
        // https://github.com/apache/arrow-datafusion/pull/6629
        schemas.sort_by(|(location1, _), (location2, _)| location1.cmp(location2));
```

**Review by alamb — APPROVED** [review]
Thank you @thomas-k-cameron  -- I think this looks great. The only potential downside of this change I can think of is that this will change behavior for any object store that lists files sorted deterministicially in a different order than `sort()` 

However, I think most object stores sort their locations lexographically so this PR is likely to simply make local filesystems consistent. 

Thanks again

**Comment by alamb** [comment]
Thanks @thomas-k-cameron  -- once the check pass I'll plan to merge this PR.

**Comment by alamb** [comment]
🚀

**Pull request merged**